src/Features/Core/Portable/RemoveUnusedMembers/AvoidUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.AvoidUnusedMembers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class AvoidUnusedMembersDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        internal static string UnunsedMemberNameProperty = nameof(UnunsedMemberNameProperty);
+        internal static string UnunsedMemberKindProperty = nameof(UnunsedMemberKindProperty);
+
+        private readonly DiagnosticDescriptor _noReadsUnnecessaryWithSuggestionDescriptor;
+
+        public AvoidUnusedMembersDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.AvoidUnusedMembersDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Title), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Unused_Members_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+            _noReadsUnnecessaryWithSuggestionDescriptor = new DiagnosticDescriptor(
+                UnnecessaryWithSuggestionDescriptor.Id,
+                UnnecessaryWithSuggestionDescriptor.Title,
+                new LocalizableResourceString(nameof(FeaturesResources.Avoid_Members_Without_Reads_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
```

Comment:
this feels super weird... same diagnostic ID... but different message....

Reply:
This is a very common pattern in FxCop analyzers where we have large number of minor variations for the same underlying issue, here "an unused member". The pattern we chose there was to avoid assigning a new ID for each such case, and instead just have different descriptors with differing messages. This also improves the FixAll experience by avoiding extra iterations per ID. I am not sure what is the pattern for such scenarios in the IDE analyzers, if we have encountered one till now. I can merge both the cases into a single message, but I felt it is helpful to provide the distinction to the end user, especially as we already know it in the analyzer.

---
In reply to: [212729488](https://github.com/dotnet/roslyn/pull/29511#discussion_r212729488) [](ancestors = 212729488)

Reply:
Well, is there a problem with just havin a suitable single message?  The specific code-fix can have a particular message to show the user if necessary.

After all, you only have one option controlling this.  And you have fix-all fix all of these up.  So distinguishing them in any way at all seems superfluous to me.

Reply:
> I am not sure what is the pattern for such scenarios in the IDE analyzers

IDE analyzer pattern is that these are either actually different things that should be displayed to the users, or they're not.  If htey are, then they should be controlled through separate mechanisms (like options) and should not participate in fix-all together (since we have presented them as being different).  If they aren't considered different things (i.e. you can only turn them all on/off at once, and you fix them all together), then they have the same message.

Reply:
I have taken this route - we can split into multiple messages/options/IDs if needed in future.

---
In reply to: [212774669](https://github.com/dotnet/roslyn/pull/29511#discussion_r212774669) [](ancestors = 212774669)

Reply:
Sounds good.  Note, if you want, you can always have different messages for the *fix* title.  That part is very flexible and can be different if you think it would provide clarity and whatnot.  For example, you could have 'Remove unused member' as well as 'Remove unwritten member'.